spec_util/meld.go

```diff
@@ -271,9 +271,76 @@ func haveCompatibleTypes(dst, src *pb.Primitive) bool {
 		return false
 	}
 
-	cmpDst := &pb.Primitive{Value: dst.Value, FormatKind: dst.FormatKind}
-	cmpSrc := &pb.Primitive{Value: src.Value, FormatKind: src.FormatKind}
-	return proto.Equal(cmpDst, cmpSrc)
+	// First, check that the base types can join.
+	baseJoin := joinBaseTypes(dst, src)
+	if baseJoin == nil {
+		return false
+	}
+
+	// Types are compatible if the base types can join and the format kinds
+	// are equal.
+	cmpDst := &pb.Primitive{Value: baseJoin.Value, FormatKind: dst.FormatKind}
+	cmpSrc := &pb.Primitive{Value: baseJoin.Value, FormatKind: src.FormatKind}
+	if proto.Equal(cmpDst, cmpSrc) {
+		return true
+	}
+
+	// Types are compatible if the base types can join and least one type does
+	// not have any data formats identified.
+	if dst.FormatKind == "" || src.FormatKind == "" {
+		return true
+	}
+
+	return false
+}
+
+// Returns a new Primitive with the Value set as the type-theoretic join of
+// dst.Value and src.Value.  For example, join(int32, uint32) = int64.
+// Returns nil if no such join exists, e.g. join(int64, uint64) = nil.
+func joinBaseTypes(dst, src *pb.Primitive) *pb.Primitive {
+	dstType := reflect.TypeOf(dst.Value)
+	srcType := reflect.TypeOf(src.Value)
+
+	if dstType == srcType {
+		return &pb.Primitive{Value: dst.Value}
+	}
+
+	switch dst.Value.(type) {
+	case *pb.Primitive_Int32Value:
+		switch src.Value.(type) {
+		case *pb.Primitive_Int64Value, *pb.Primitive_Uint32Value:
+			return &pb.Primitive{Value: &pb.Primitive_Int64Value{Int64Value: &pb.Int64{}}}
+		}
+	case *pb.Primitive_Int64Value:
+		switch src.Value.(type) {
+		case *pb.Primitive_Int32Value, *pb.Primitive_Uint32Value:
+			return &pb.Primitive{Value: &pb.Primitive_Int64Value{Int64Value: &pb.Int64{}}}
+		}
+	case *pb.Primitive_Uint32Value:
+		switch src.Value.(type) {
+		case *pb.Primitive_Int32Value:
+			return &pb.Primitive{Value: &pb.Primitive_Int64Value{Int64Value: &pb.Int64{}}}
+		case *pb.Primitive_Uint64Value:
+			return &pb.Primitive{Value: &pb.Primitive_Uint64Value{Uint64Value: &pb.Uint64{}}}
+		}
+	case *pb.Primitive_Uint64Value:
+		switch src.Value.(type) {
+		case *pb.Primitive_Uint32Value:
+			return &pb.Primitive{Value: &pb.Primitive_Uint64Value{Uint64Value: &pb.Uint64{}}}
+		}
```

Comment:
This section looks good.

I was a bit concerned that we can't join Int64 with UInt64.  So I checked the parsing code, which for JSON numbers tries Int64 first, then Int64 again, then Uint64, then float, then bool.  (CategorizeString is called only after json.Number.Int64() fails.)

So now I'm a bit confused about how we get 32-bit integers at all?  Or don't we?  (For uploaded specs?)

It might be worth commenting above that our assumption while parsing is that everything that can fit into a signed int64 is of that type, so uint64 is really only used for values >= 2^63.

Reply:
Sounds good, I added the comment.

What was your concern about not being able to join Int64 and UInt64?  That's what I had expected -- to join, we'd need a signed integer type large enough to represent all values in Int64 and UInt64, which doesn't exist in our (admittedly limited) type system.

Reply:
I wasn't sure when we decided something was an int64 vs. an uint64.  The answer appears to be that we always assume it's signed, unless it's too large for that.  But not being able to merge -5000000000 with +5000000000 because they are a signed and unsigned int64 would have been a problem.